src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -59,37 +83,71 @@ private bool GetNext()
                 return false;
             }
 
+            internal abstract Set<TSource> FillSet();
+
+            public TSource[] ToArray()
+            {
+                return FillSet().ToArray();
+            }
+
+            public List<TSource> ToList()
+            {
+                return FillSet().ToList();
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                return onlyIfCheap ? -1 : FillSet().Count;
+            }
+        }
+
+        private sealed class UnionIterator2<TSource> : UnionIterator<TSource>
+        {
+            private readonly IEnumerable<TSource> _first;
+            private readonly IEnumerable<TSource> _second;
+
+            public UnionIterator2(IEnumerable<TSource> first, IEnumerable<TSource> second, IEqualityComparer<TSource> comparer)
+                : base(comparer)
+            {
+                Debug.Assert(first != null);
+                Debug.Assert(second != null);
+                _first = first;
+                _second = second;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new UnionIterator2<TSource>(_first, _second, Comparer);
+            }
+
             public override bool MoveNext()
             {
-                switch(state)
+                switch (state)
```

Comment:
Wny do these need specialized MoveNext methods?  Why not just have one in the base as in Concat?  I'd much prefer to have the single implementation, unless there's a very measurable perf difference... is there?